### ISSUE TITLE
Extension loaded starts after spock-spring extension

### ISSUE THREAD

**Issue by czyzniek** [joke/spock-mockable#9]
I tried to use `spock-mockable` in my PoC project where I have Kotlin + spring-boot. I have sample configuration class where I want to create few mocks:
[SpockWebContractBaseClass](https://github.com/czyzniek/bank/blob/with-spock/accounting/src/test/groovy/pl/sii/bank/accounting/SpockWebContractBaseClass.groovy)
 Unfortunately I end up with following exception:
 ```
 Unable to redefined classes: [class pl.sii.bank.accounting.domain.CreateAccountForCustomerUseCase, class pl.sii.bank.accounting.domain.FetchAllCustomersUseCase, class pl.sii.bank.accounting.domain.CreateCustomerUserCase, class pl.sii.bank.accounting.domain.FetchAccountBalanceUseCase].
org.spockframework.runtime.SpockExecutionException: Unable to redefined classes: [class pl.sii.bank.accounting.domain.CreateAccountForCustomerUseCase, class pl.sii.bank.accounting.domain.FetchAllCustomersUseCase, class pl.sii.bank.accounting.domain.CreateCustomerUserCase, class pl.sii.bank.accounting.domain.FetchAccountBalanceUseCase].
	at io.github.joke.spockmockable.internal.MockableExtension$InstallationListener.onError(MockableExtension.groovy:91)
	at net.bytebuddy.agent.builder.AgentBuilder$InstallationListener$Compound.onError(AgentBuilder.java:4477)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.doInstall(AgentBuilder.java:9515)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:9423)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Delegator.installOn(AgentBuilder.java:11025)
	at io.github.joke.spockmockable.internal.MockableExtension.installTransformation(MockableExtension.groovy:58)
	at io.github.joke.spockmockable.internal.MockableExtension.visitSpecAnnotation(MockableExtension.groovy:48)
	at io.github.joke.spockmockable.internal.MockableExtension.visitSpecAnnotation(MockableExtension.groovy)
	at org.spockframework.runtime.ExtensionRunner.doRunAnnotationDrivenExtensions(ExtensionRunner.java:91)
	at org.spockframework.runtime.ExtensionRunner.runAnnotationDrivenExtensions(ExtensionRunner.java:62)
	at org.spockframework.runtime.ExtensionRunner.runAnnotationDrivenExtensions(ExtensionRunner.java:60)
	at org.spockframework.runtime.ExtensionRunner.runAnnotationDrivenExtensions(ExtensionRunner.java:51)
	at org.spockframework.runtime.ExtensionRunner.run(ExtensionRunner.java:41)
	at org.spockframework.runtime.Sputnik.runExtensionsIfNecessary(Sputnik.java:90)
	at org.spockframework.runtime.Sputnik.getDescription(Sputnik.java:54)
	at org.junit.runners.model.RunnerBuilder.configureRunner(RunnerBuilder.java:81)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:72)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:37)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:70)
	at org.junit.internal.requests.ClassRequest.createRunner(ClassRequest.java:28)
	at org.junit.internal.requests.MemoizingRequest.getRunner(MemoizingRequest.java:19)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:78)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.github.joke.spockmockable.internal.MockableExtension$ClassRedefinitionException
	at io.github.joke.spockmockable.internal.MockableExtension$RedefinitionListener.onComplete(MockableExtension.groovy:82)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Listener$Compound.onComplete(AgentBuilder.java:5711)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector.apply(AgentBuilder.java:7039)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy.apply(AgentBuilder.java:4868)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.doInstall(AgentBuilder.java:9502)
	... 48 more
 ```
 Could you tell me what I am doing wrong and why spock-mockable is not able to unfinal those classes?

**Comment by joke** [comment]
I never tried it with kotlin. My guess is the classes are loaded before spock-mockable gets a chance to redefine them.
I'll take a look alter on. Thanks for providing a demo. That really helps.

**Comment by joke** [comment]
Your gradle setup seems rather strange. It's not possible to build the project on branch `with-spock`.
````
$ ./gradlew build

> Task :buildEnvironment

------------------------------------------------------------
Root project
------------------------------------------------------------

classpath
No dependencies

A web-based, searchable dependency report is available by adding the --scan option.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 304ms
1 actionable task: 1 executed
````

**Comment by czyzniek** [comment]
This is not multi-module Gradle project. Please go to `accounting` directory and run this command one more time ;)

**Comment by joke** [comment]
I added a kotlin test. It kotlin shot be supported in general: https://github.com/joke/spock-mockable/tree/main/examples/kotlin

**Comment by joke** [comment]
Seems the `spock-spring` extensions initializes the context before `spock-mockable` gets a chance to register the the class redefinition. And since the context initializes the class `FetchAccountBalanceUseCase` the class manipulation will fail.

**Comment by czyzniek** [comment]
Is there a way to force load order of these extensions?

**Comment by joke** [comment]
Currently not. The problem is a bit more complicated. I'm looking for ways fix the issue.
I'll let you know if I'm able to fix it.

**Comment by joke** [comment]
The issue should be fixed. But I did not try it which your protect. Artifacts in maven central will be available soon.

**Comment by czyzniek** [comment]
I can confirm, it works! Thanks for help :)

**Comment by leonard84** [comment]
@joke I'd suggest to attach your agent in the static initializer instead of waiting for spock to call `start()`.

**Comment by joke** [comment]
> @joke I'd suggest to attach your agent in the static initializer instead of waiting for spock to call `start()`.

Good idea. Thanks

**Comment by kczyznielewski-sii** [comment]
Additionally, I noticed strange behavior. When we declare `abstract` class that extends `Specification` and in this class we use `@Mockable` we are unable to create mocks:
```
    org.spockframework.mock.CannotCreateMockException at JavaMockFactory.java:48
    java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
        Caused by: java.lang.IllegalArgumentException at TypeCache.java:155
            Caused by: java.lang.IllegalArgumentException at ByteBuddy.java:406
```
You can check my master branch and change `pl.sii.bank.accounting.WebContractBaseSpec` for abstract class.

**Comment by joke** [comment]
Hi,
I just released version 1.3.0 which contains a lot of changes and rewrites.
At least in that version inheritance should be working: [ControllerTestFromBase.groovy](https://github.com/joke/spock-mockable/blob/main/examples/spring/src/test/groovy/io/github/joke/spockmockable/ControllerTestFromBase.groovy).

I'll take a look at your project later on. In the meantime you could try the new release. The new version should be available in maven central soon.

**Comment by joke** [comment]
@czyzniek If you run your `WebSpec` manually everything works.
The problem is your `MessagingSpec` is the usecases as well. You need to open the classes as well or start the spec in an individual jvm by setting test { forkEvery 1 } in gradle.

**Comment by czyzniek** [comment]
Yeah, I gave a try. The funny part is when I remove `abstract` from base class, everything works. There is no runtime exception, nothing. It looks like `abstract` in my case is the culprit :/

**Comment by joke** [comment]
It's not the abstract. If your remove the abstract the test execution order changes the class it opened before `MessagingSpec` get's to run.
Your problem inspired me to changed to way how the redefinition is triggered. Might will around with it and if it works I'll release again.

**Comment by czyzniek** [comment]
Man, you are making my day! Looking forward to this ;)

**Comment by joke** [comment]
@czyzniek I tested your problem with a more condensed example. I seems to be working with [`Spock 2.0-M4`](http://spockframework.org/spock/docs/2.0-M4/all_in_one.html) and `spock-mockable 1.3.0` out of the box. The way how spock internally loads classes seems to be different and thus the `@Mockable` annotation is picked out earlier.

**Comment by leonard84** [comment]
@joke your extension is similar to what I brainstormed myself. However, instead of writing it all in annotations that might be encountered sometime, I thought about creating a simple file in `META-INF/` that contains all classes to be un-finaled, this way it would be more consistent as you'd know about all classes ahead of time.

**Comment by joke** [comment]
@leonard84 I though of using a file in META-INF myself but I voted against initially.

Currently a user can disable the loading process if he want's to
  * By starting everything in a seperat JVM and test the unmodified classes if he likes.
  * Or add different "tests cycles" based on file pattern or whatever by using a new gradle task or maven goal.

But based on the discussion with @czyzniek im currently thinking of supporting both ways with an option on the annotation:

@Mockable(classes = Person, scope = GLOBAL)
@Mockable(classes = Person, scope = TEST)
class PersonTest extends Specification {}

I'm planning on supporting something to open classes based on packages but I didn't have the time to do it just yet.

**Comment by leonard84** [comment]
Personally, I think it is best to keep it simple so global should be the default. Otherwise, you might run into issues with having to redefine already loaded classes, since another test ran first. If you want to opt out of the agent, I'd prefer to start the test with a SystemProperty `mockable.enabled=false` or something like it. As for `scope=TEST` keep in mind that with 2.0-M4+ Spock supports parallel execution and since classes are global, you'll probably run into issues.

Are there reasons, to not globally un-finalizing those classes, e.g. will code break if something is not final? Or, is it just a feeling of not wanting to affect to much?

**Comment by joke** [comment]
@leonard84 the SystemProperties is indeed a good option. I agree with your point of keeping it simple.
Basically it's just a feeling. But there are some rare cases were thinks might actually break.

Come to think of it currently my main concern is:
If you annotate one test and and you open the class for the entire JVM and other tests will magically work even if you forget to annotate the test that would require it. This happens if you run `gradle test` for example. 
But if you manually run a single test that is missing the annotation from your IDE or via gradle test options at least the test will fail. That's what you expect since the `@Mockable` is missing.
But if you do it via META-INF or something like that the single test would be running without a problem even though it's not annotated correctly. And it get's a little worse if you delete the test with `@Mockable` all you will fail because way were missing the missing in the first place but you never noticed before.

Haven't though of a good solution so far.

**Comment by leonard84** [comment]
Yes, it is a tricky one. Although, it is not terribly hard to figure out why the other test starts failing.

I would treat the annotation as a nice, refactor safe version of maintaining this file. I mean, you could also have a single dummy file that just collects all the classes to unlock and thus manage it in a central location.

**Comment by leonard84** [comment]
Just FYI you could even use JUnits `LauncherSessionListener` mentioned in https://github.com/junit-team/junit5/issues/2429 to run as early as possible. If you go that route and the meta-inf file, then this would actually work for any JUnit Platform based TestEngine.

**Comment by joke** [comment]
Just released version 1.4.0. I went with the META-INF approach and start transforming classes as early as possible. 

@leonard84 Thanks for the suggestions
